src/core/componentSystems/MaskingSystem.js

```diff
@@ -66,82 +134,88 @@ export class MaskingSystem extends MRSystem {
         if (entity instanceof MRPanel) {
-            // Using an array for the panels in case we need them for more manipulations down the line instead
-            // of using the system's registry.
-            this.panels.push(entity);
-
-            // Need to set stencilRef for the children of this panel to match that of this panel so
-            // that when rendering the children they only mask based on the panel's geometry location instead
-            // of all panel geometry locations.
-            //
-            // stencilRef needs to be > 0 as 0 is the webgl default and -1 is our manual default of 'not set yet'.
-            // We're basing our stencilRef on the 1+index location (ie length of array at adding) of the panel entity.
-            // Even though we're not manually using this stencilRef in the render loop, threejs handles its use
-            // internally.
-            const stencilRef = this.panels.length;
-
-            // Currently this setup will not be able to handle properly if there is a panel within another
-            // panel in the html setup. Defaulting that case to be based on whichever panel is the entity
-            // passed through this function since that case is an edge case that will not be expected.
-
-            /**
-             *
-             * @param child
-             */
-            function runTheTraversal(child) {
+            if (this.panelCount >= MAX_PANEL_NUM) {
+                console.warn('Masking system supports up to eight panels.');
+                return;
+            }
+
+            // Ignoring panel removal for now.
+            // TODO: Handle panel removal
+            const stencilRefShift = this.panelCount;
+            this.panelCount++;
+
+            entity.traverse((child) => {
                 if (child instanceof MRPanel && child.object3D.isGroup) {
                     // The panel entity should contain a group object where the first panel child we hit is this panel itself.
                     // We need to mask based off the background mesh of this object.
-                    let mesh = child.background;
-                    if (this.app.debug && mesh.material.color) {
-                        mesh.material.color.set(0xff00ff); // pink
-                    }
-                    mesh.material.stencilWrite = this.panelStencilMaterial.stencilWrite;
-                    mesh.material.stencilFunc = this.panelStencilMaterial.stencilFunc;
-                    mesh.material.stencilRef = stencilRef;
-                    mesh.material.stencilZPass = this.panelStencilMaterial.stencilZPass;
+                    const sourceObj = child.background;
+
+                    // TODO: Optimize material.
+                    // Since only needs to write to the stencil buffer, no need to write to the color buffer,
+                    // therefore, we can use a simpler material than MeshBasicMaterial. Should we use
+                    // ShaderMaterial?
```

Comment:
I think switching to ShaderMaterial would be good but if I remember correctly when I've done that before in mrjs it can hit a weird edge case for some aspect of physics our bounding box updates (i forget exactly where)

just noting this here for future if that ends up being something you'd want to chase down

not blocker or priority for this pr though

Reply:
I don't think material is involved for bounding box but yes I agree with that we might encounter some (Raw)ShaderMaterial specific edge cases if we use it. Perhaps using regular materials would be safer. And I don't think using Basic material here can cause big bad performance impact because just up to 8 panels (rects?) are rendered under the current limitation. This inline comment is for future advanced optimization.